### PR TITLE
feat(examples): add FSDP

### PR DESCRIPTION
**Pull request by tpoisonooo**
# Content

Add an FSDP example for `Muon` optimizer.

Here is the difference:

1. Add `DistributedSampler` 
```python
    sampler = DistributedSampler(dataset=train_dataset,
                                 rank=rank,
                                 num_replicas=world_size,
                                 shuffle=True)
```
2. Add all reduce
```python
dist.all_reduce(loss, op=dist.ReduceOp.AVG)
```

# Test

loss:
image

GPU usage:

image

I am also a FSDP begginer, for any question, I would try to fix.  #25

### CODE REVIEW

**Comment by tpoisonooo** [comment]
cc @toothacher17

**Comment by tpoisonooo** [comment]
Furthermore, `toy_run_fsdp.py` is compatible with `toy_run.dy`,  for single GPU, just:

```bash
export CUDA_VISIBLE_DEVICES="0"
python3 toy_run_fsdp.py
```

They are same.
